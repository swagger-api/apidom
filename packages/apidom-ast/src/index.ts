// JSON AST related exports
export { default as JsonNode } from './json/nodes/JsonNode.ts';
export { default as JsonDocument } from './json/nodes/JsonDocument.ts';
export { default as JsonObject } from './json/nodes/JsonObject.ts';
export { default as JsonProperty } from './json/nodes/JsonProperty.ts';
export type { JsonValue as JsonPropertyValue } from './json/nodes/JsonProperty.ts';
export { default as JsonArray } from './json/nodes/JsonArray.ts';
export { default as JsonValue } from './json/nodes/JsonValue.ts';
export type { JsonValueOptions } from './json/nodes/JsonValue.ts';
export { default as JsonKey } from './json/nodes/JsonKey.ts';
export { default as JsonString } from './json/nodes/JsonString.ts';
export { default as JsonStringContent } from './json/nodes/JsonStringContent.ts';
export { default as JsonEscapeSequence } from './json/nodes/JsonEscapeSequence.ts';
export { default as JsonNumber } from './json/nodes/JsonNumber.ts';
export { default as JsonTrue } from './json/nodes/JsonTrue.ts';
export { default as JsonFalse } from './json/nodes/JsonFalse.ts';
export { default as JsonNull } from './json/nodes/JsonNull.ts';
export {
  isDocument as isJsonDocument,
  isFalse as isJsonFalse,
  isProperty as isJsonProperty,
  isStringContent as isJsonStringContent,
  isEscapeSequence as isJsonEscapeSequence,
  isArray as isJsonArray,
  isKey as isJsonKey,
  isNull as isJsonNull,
  isNumber as isJsonNumber,
  isObject as isJsonObject,
  isString as isJsonString,
  isTrue as isJsonTrue,
} from './json/nodes/predicates.ts';
// YAML AST related exports
export { default as YamlAlias } from './yaml/nodes/YamlAlias.ts';
export type { YamlAliasOptions } from './yaml/nodes/YamlAlias.ts';
export { default as YamlCollection } from './yaml/nodes/YamlCollection.ts';
export { default as YamlComment } from './yaml/nodes/YamlComment.ts';
export type { YamlCommentOptions } from './yaml/nodes/YamlComment.ts';
export { default as YamlDirective } from './yaml/nodes/YamlDirective.ts';
export type { YamlDirectiveOptions, YamlDirectiveParameters } from './yaml/nodes/YamlDirective.ts';
export { default as YamlDocument } from './yaml/nodes/YamlDocument.ts';
export { default as YamlKeyValuePair } from './yaml/nodes/YamlKeyValuePair.ts';
export type { YamlKeyValuePairOptions } from './yaml/nodes/YamlKeyValuePair.ts';
export { default as YamlMapping } from './yaml/nodes/YamlMapping.ts';
export { default as YamlNode } from './yaml/nodes/YamlNode.ts';
export type { YamlNodeOptions } from './yaml/nodes/YamlNode.ts';
export { default as YamlScalar } from './yaml/nodes/YamlScalar.ts';
export type { YamlScalarOptions } from './yaml/nodes/YamlScalar.ts';
export { default as YamlSequence } from './yaml/nodes/YamlSequence.ts';
export { default as YamlStream } from './yaml/nodes/YamlStream.ts';
export { default as YamlTag, YamlNodeKind } from './yaml/nodes/YamlTag.ts';
export type { YamlTagOptions } from './yaml/nodes/YamlTag.ts';
export { default as YamlAnchor } from './yaml/nodes/YamlAnchor.ts';
export type { YamlAnchorOptions } from './yaml/nodes/YamlAnchor.ts';
export { YamlStyle, YamlStyleGroup } from './yaml/nodes/YamlStyle.ts';
export { default as YamlFailsafeSchema } from './yaml/schemas/failsafe/index.ts';
export { default as YamlJsonSchema } from './yaml/schemas/json/index.ts';
export { default as YamlReferenceManager } from './yaml/anchors-aliases/ReferenceManager.ts';
export {
  isAlias as isYamlAlias,
  isAnchor as isYamlAnchor,
  isComment as isYamlComment,
  isKeyValuePair as isYamlKeyValuePair,
  isDirective as isYamlDirective,
  isDocument as isYamlDocument,
  isMapping as isYamlMapping,
  isScalar as isYamlScalar,
  isSequence as isYamlSequence,
  isStream as isYamlStream,
  isTag as isYamlTag,
} from './yaml/nodes/predicates.ts';
export { default as YamlError } from './yaml/errors/YamlError.ts';
export { default as YamlReferenceError } from './yaml/errors/YamlReferenceError.ts';
export { default as YamlSchemaError } from './yaml/errors/YamlSchemaError.ts';
export { default as YamlTagError } from './yaml/errors/YamlTagError.ts';
export type { YamlTagErrorOptions } from './yaml/errors/YamlTagError.ts';
// generic AST related exports
export type { default as Node, NodeOptions } from './Node.ts';
export { default as Literal } from './Literal.ts';
export type { LiteralOptions } from './Literal.ts';
export { Point, default as Position } from './Position.ts';
export type { PointOptions, PositionOptions } from './Position.ts';
export { default as Error } from './Error.ts';
export type { ErrorOptions } from './Error.ts';
export { default as ParseResult } from './ParseResult.ts';
export { isParseResult, isLiteral, isPoint, isPosition } from './predicates.ts';
// AST traversal related exports
export {
  BREAK,
  mergeAll as mergeAllVisitors,
  getVisitFn,
  visit,
  getNodeType,
  isNode,
  cloneNode,
} from './traversal/visitor.ts';
export type { MergeAllSync, MergeAllAsync } from './traversal/visitor.ts';
