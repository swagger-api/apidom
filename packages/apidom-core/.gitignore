/dist
/es
/cjs
/types
/NOTICE
/swagger-api-apidom-core-*.tgz
