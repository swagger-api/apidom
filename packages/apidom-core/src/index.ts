import { has } from 'ramda';
import { isPlainObject, isString } from 'ramda-adjunct';
import { Element, Namespace as INamespace } from 'minim';

import defaultNamespaceInstance from './namespace';
import serializeValue from './serializers/value';
import serializeJSON from './serializers/json';
import serializeYAML from './serializers/yaml-1-2';

export { dispatchPlugins as dispatchRefractorPlugins } from './refractor/plugins/utils';
export { default as refractorPluginElementIdentity } from './refractor/plugins/element-identity';
export { default as refractorPluginSemanticElementIdentity } from './refractor/plugins/semantic-element-identity';

export { default as MediaTypes } from './media-types';

export { Element, MemberElement, KeyValuePair, ObjectSlice, ArraySlice, refract } from 'minim';
export type { NamespacePluginOptions, Attributes, Meta } from 'minim';
export type { PositionRange, Position } from './elements/SourceMap';
export { default as namespace, Namespace, createNamespace } from './namespace';

export {
  ObjectElement,
  ArrayElement,
  BooleanElement,
  NullElement,
  NumberElement,
  StringElement,
  LinkElement,
  RefElement,
  AnnotationElement,
  CommentElement,
  ParseResultElement,
  SourceMapElement,
} from './refractor/registration';

export {
  isElement,
  isStringElement,
  isNumberElement,
  isNullElement,
  isBooleanElement,
  isArrayElement,
  isObjectElement,
  isMemberElement,
  isLinkElement,
  isRefElement,
  isAnnotationElement,
  isParseResultElement,
  isSourceMapElement,
  isPrimitiveElement,
  hasElementSourceMap,
  includesSymbols,
  includesClasses,
} from './predicates/index';
export { default as createPredicate } from './predicates/helpers';

export { filter, reject, find, findAtOffset, some, traverse, parents } from './traversal/index';
export {
  visit,
  BREAK,
  mergeAllVisitors,
  getNodeType,
  keyMapDefault as keyMap,
} from './traversal/visitor';
export { transclude, default as Transcluder } from './transcluder/index';
export { dereference } from './util';
export { cloneShallow, cloneDeep } from './clone';
export { default as CloneError } from './clone/errors/CloneError';
export { default as DeepCloneError } from './clone/errors/DeepCloneError';
export { default as ShallowCloneError } from './clone/errors/ShallowCloneError';

/**
 * Transforms data to an Element from a particular namespace.
 */
export const from = (data: any, namespace: INamespace = defaultNamespaceInstance): Element => {
  if (isString(data)) {
    // JSON serialized refract
    try {
      return namespace.fromRefract(JSON.parse(data));
    } catch {
      // noop
    }
  }
  if (isPlainObject(data) && has('element', data)) {
    // refract javascript structure
    return namespace.fromRefract(data);
  }

  return namespace.toElement(data);
};

/**
 * Transforms the ApiDOM into JavaScript POJO.
 * This POJO would be the result of interpreting the ApiDOM
 * into JavaScript structure.
 */
export const toValue = serializeValue;

/**
 * Transforms the ApiDOM into JSON string.
 */
export const toJSON = serializeJSON;

/**
 * Transforms the ApiDOM into YAML string.
 */
export const toYAML = serializeYAML;

/**
 * Creates a refract representation of an Element.
 * https://github.com/refractproject/refract-spec
 */
export const dehydrate = (
  element: Element,
  namespace: INamespace = defaultNamespaceInstance,
): any => {
  return namespace.toRefract(element);
};

/**
 * Create a refracted string representation of an Element.
 */
export const toString = (
  element: Element,
  namespace: INamespace = defaultNamespaceInstance,
): string => {
  const refractStructure = dehydrate(element, namespace);
  return JSON.stringify(refractStructure);
};

export { default as sexprs } from './sexprs';
export { default as deepmerge } from './deepmerge';
