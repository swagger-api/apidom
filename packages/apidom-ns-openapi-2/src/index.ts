export {
  isRefElement,
  isLinkElement as isLinkPrimitiveElement,
  isMemberElement,
  isObjectElement,
  isArrayElement,
  isBooleanElement,
  isNullElement,
  isElement,
  isNumberElement,
  isStringElement,
} from '@swagger-api/apidom-core';
export {
  isJSONReferenceElement,
  JSONReferenceElement,
} from '@swagger-api/apidom-ns-json-schema-draft-4';

export { default as mediaTypes, OpenAPIMediaTypes } from './media-types';

// eslint-disable-next-line no-restricted-exports
export { default } from './namespace';

export { default as refractorPluginReplaceEmptyElement } from './refractor/plugins/replace-empty-element';

export { default as refract, createRefractor } from './refractor';
export { default as specificationObj } from './refractor/specification';

export {
  isSwaggerElement,
  isSwaggerVersionElement,
  isInfoElement,
  isContactElement,
  isLicenseElement,
  isPathsElement,
  isPathItemElement,
  isOperationElement,
  isExternalDocumentationElement,
  isParameterElement,
  isItemsElement,
  isResponsesElement,
  isResponseElement,
  isHeadersElement,
  isExampleElement,
  isHeaderElement,
  isTagElement,
  isReferenceElement,
  isSchemaElement,
  isXmlElement,
  isDefinitionsElement,
  isParametersDefinitionsElement,
  isResponsesDefinitionsElement,
  isSecurityDefinitionsElement,
  isSecuritySchemeElement,
  isScopesElement,
  isSecurityRequirementElement,
} from './predicates';

export { isReferenceLikeElement, isSwaggerExtension } from './refractor/predicates';

export { default as AlternatingVisitor } from './refractor/visitors/generics/AlternatingVisitor';
export { default as FixedFieldsVisitor } from './refractor/visitors/generics/FixedFieldsVisitor';
export { default as MapVisitor } from './refractor/visitors/generics/MapVisitor';
export { default as MixedFieldsVisitor } from './refractor/visitors/generics/MixedFieldsVisitor';
export { default as PatternedFieldsVisitor } from './refractor/visitors/generics/PatternedFieldsVisitor';
export { default as FallbackVisitor } from './refractor/visitors/FallbackVisitor';
export { default as SpecificationVisitor } from './refractor/visitors/SpecificationVisitor';
export { default as Visitor } from './refractor/visitors/Visitor';

export { keyMap, getNodeType } from './traversal/visitor';

// OpenAPI 2.0 elements
export {
  SwaggerElement,
  SwaggerVersionElement,
  InfoElement,
  ContactElement,
  LicenseElement,
  PathsElement,
  PathItemElement,
  OperationElement,
  ExternalDocumentationElement,
  ParameterElement,
  ItemsElement,
  ResponsesElement,
  ResponseElement,
  HeadersElement,
  ExampleElement,
  HeaderElement,
  TagElement,
  ReferenceElement,
  SchemaElement,
  XmlElement,
  DefinitionsElement,
  ParametersDefinitionsElement,
  ResponsesDefinitionsElement,
  SecurityDefinitionsElement,
  SecuritySchemeElement,
  ScopesElement,
  SecurityRequirementElement,
} from './refractor/registration';
// NCE types
export { default as OperationConsumesElement } from './elements/nces/OperationConsumes';
export { default as OperationParametersElement } from './elements/nces/OperationParameters';
export { default as OperationProducesElement } from './elements/nces/OperationProduces';
export { default as OperationSchemesElement } from './elements/nces/OperationSchemes';
export { default as OperationSecurityElement } from './elements/nces/OperationSecurity';
export { default as OperationTagsElement } from './elements/nces/OperationTags';
export { default as PathItemParametersElement } from './elements/nces/PathItemParameters';
export { default as SwaggerSchemesElement } from './elements/nces/SwaggerSchemes';
export { default as SwaggerConsumesElement } from './elements/nces/SwaggerConsumes';
export { default as SwaggerProducesElement } from './elements/nces/SwaggerProduces';
export { default as SwaggerSecurityElement } from './elements/nces/SwaggerSecurity';
export { default as SwaggerTagsElement } from './elements/nces/SwaggerTags';
