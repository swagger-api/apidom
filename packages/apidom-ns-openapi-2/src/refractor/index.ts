import { invokeArgs } from 'ramda-adjunct';
import {
  visit,
  Element,
  dereference,
  refract as baseRefract,
  dispatchRefractorPlugins,
} from '@swagger-api/apidom-core';

import specification from './specification';
import { keyMap, getNodeType } from '../traversal/visitor';
import createToolbox from './toolbox';

const refract = <T extends Element>(
  value: any,
  { specPath = ['visitors', 'document', 'objects', 'Swagger', '$visitor'], plugins = [] } = {},
): T => {
  const element = baseRefract(value);
  const resolvedSpec = dereference(specification);

  /**
   * This is where generic ApiDOM becomes semantic (namespace applied).
   * We don't allow consumers to hook into this translation.
   * Though we allow consumers to define their onw plugins on already transformed ApiDOM.
   */
  const rootVisitor = invokeArgs(specPath, [], resolvedSpec);
  // @ts-ignore
  visit(element, rootVisitor, { state: { specObj: resolvedSpec } });

  /**
   * Running plugins visitors means extra single traversal === performance hit.
   */
  return dispatchRefractorPlugins(rootVisitor.element, plugins, {
    toolboxCreator: createToolbox,
    visitorOptions: { keyMap, nodeTypeGetter: getNodeType },
  });
};

export const createRefractor =
  (specPath: string[]) =>
  (value: any, options = {}) =>
    refract(value, { specPath, ...options });

export default refract;
