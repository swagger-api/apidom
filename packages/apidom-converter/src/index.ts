import { ParseResultElement } from '@swagger-api/apidom-core';
import { mergeOptions, bundle, File } from '@swagger-api/apidom-reference';

import defaultOptions, { ConverterOptions } from './options';
import ConvertError from './errors/ConvertError';
import UnmatchedConvertStrategyError from './errors/UnmatchedConvertStrategyError';

export { ConvertError, UnmatchedConvertStrategyError };

/**
 * `convertApiDOM` already assumes that the ApiDOM is bundled.
 */
export const convertApiDOM = async (element: ParseResultElement, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const file = File({
    uri: mergedOptions.resolve.baseURI,
    parseResult: element,
    mediaType: mergedOptions.convert.sourceMediaType || mergedOptions.parse.mediaType,
  });
  const strategy = mergedOptions.convert.strategies.find((s) => s.canConvert(file, mergedOptions));

  if (typeof strategy === 'undefined') {
    throw new UnmatchedConvertStrategyError(file.uri);
  }

  return strategy.convert(file, mergedOptions);
};

const convert = async (uri: string, options = {}) => {
  const mergedOptions = mergeOptions(defaultOptions, options || {}) as ConverterOptions;
  const parseResult = await bundle(uri, mergedOptions);

  return convertApiDOM(parseResult, mergedOptions);
};

export default convert;
