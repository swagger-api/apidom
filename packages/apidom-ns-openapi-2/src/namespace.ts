import { NamespacePluginOptions } from '@swagger-api/apidom-core';

import InfoElement from './elements/Info';
import LicenseElement from './elements/License';
import ContactElement from './elements/Contact';
import ExternalDocumentation from './elements/ExternalDocumentation';
import ItemsElement from './elements/Items';
import ExampleElement from './elements/Example';
import HeaderElement from './elements/Header';
import TagElement from './elements/Tag';
import XmlElement from './elements/Xml';
import SecurityDefinitionsElement from './elements/SecurityDefinitions';
import SecuritySchemeElement from './elements/SecurityScheme';
import ScopesElement from './elements/Scopes';
import SecurityRequirementElement from './elements/SecurityRequirement';

const openApi2 = {
  namespace: (options: NamespacePluginOptions) => {
    const { base } = options;

    base.register('info', InfoElement);
    base.register('license', LicenseElement);
    base.register('contact', ContactElement);
    base.register('externalDocumentation', ExternalDocumentation);
    base.register('items', ItemsElement);
    base.register('example', ExampleElement);
    base.register('header', HeaderElement);
    base.register('tag', TagElement);
    base.register('xml', XmlElement);
    base.register('securityDefinitions', SecurityDefinitionsElement);
    base.register('securityScheme', SecuritySchemeElement);
    base.register('scopes', ScopesElement);
    base.register('securityRequirement', SecurityRequirementElement);

    return base;
  },
};

export default openApi2;
