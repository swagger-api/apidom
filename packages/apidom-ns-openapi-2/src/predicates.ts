import { createPredicate } from '@swagger-api/apidom-core';

import InfoElement from './elements/Info';
import LicenseElement from './elements/License';
import ContactElement from './elements/Contact';
import ExternalDocumentation from './elements/ExternalDocumentation';
import XmlElement from './elements/Xml';
import SecurityDefinitionsElement from './elements/SecurityDefinitions';
import SecuritySchemeElement from './elements/SecurityScheme';
import SecurityRequirementElement from './elements/SecurityRequirement';
import ScopesElement from './elements/Scopes';

export const isInfoElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof InfoElement ||
      (hasBasicElementProps(element) &&
        isElementType('info', element) &&
        primitiveEq('object', element));
  },
);

export const isLicenseElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof LicenseElement ||
      (hasBasicElementProps(element) &&
        isElementType('license', element) &&
        primitiveEq('object', element));
  },
);

export const isContactElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ContactElement ||
      (hasBasicElementProps(element) &&
        isElementType('contact', element) &&
        primitiveEq('object', element));
  },
);
export const isExternalDocumentationElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ExternalDocumentation ||
      (hasBasicElementProps(element) &&
        isElementType('externalDocumentation', element) &&
        primitiveEq('object', element));
  },
);

export const isXmlElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof XmlElement ||
      (hasBasicElementProps(element) &&
        isElementType('xml', element) &&
        primitiveEq('object', element));
  },
);

export const isSecurityDefinitionsElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof SecurityDefinitionsElement ||
      (hasBasicElementProps(element) &&
        isElementType('securityDefinitions', element) &&
        primitiveEq('object', element));
  },
);

export const isSecuritySchemeElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof SecuritySchemeElement ||
      (hasBasicElementProps(element) &&
        isElementType('securityScheme', element) &&
        primitiveEq('object', element));
  },
);

export const isScopesElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof ScopesElement ||
      (hasBasicElementProps(element) &&
        isElementType('scopes', element) &&
        primitiveEq('object', element));
  },
);

export const isSecurityRequirementElement = createPredicate(
  ({ hasBasicElementProps, isElementType, primitiveEq }) => {
    return (element: any) =>
      element instanceof SecurityRequirementElement ||
      (hasBasicElementProps(element) &&
        isElementType('securityRequirement', element) &&
        primitiveEq('object', element));
  },
);
