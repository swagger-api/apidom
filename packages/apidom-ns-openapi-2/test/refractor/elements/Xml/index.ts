import { expect, assert } from 'chai';
import { sexprs, includesClasses } from '@swagger-api/apidom-core';

import { XmlElement } from '../../../../src';

describe('refractor', function () {
  context('elements', function () {
    context('XmlElement', function () {
      specify('should refract to semantic ApiDOM tree', function () {
        const xmlElement = XmlElement.refract({
          name: 'animal',
          namespace: 'http://swagger.io/schema/sample',
          prefix: 'sample',
          attribute: true,
          wrapped: false,
        });

        expect(sexprs(xmlElement)).toMatchSnapshot();
      });

      specify('should support specification extensions', function () {
        const xmlElement = XmlElement.refract({
          name: 'animal',
          'x-extension': 'extension',
        }) as XmlElement;

        assert.isFalse(includesClasses(['specification-extension'], xmlElement.getMember('name')));
        assert.isTrue(
          includesClasses(['specification-extension'], xmlElement.getMember('x-extension')),
        );
      });
    });
  });
});
