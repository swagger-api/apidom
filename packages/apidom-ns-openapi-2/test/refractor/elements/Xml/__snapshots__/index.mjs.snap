// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refractor elements XmlElement should refract to semantic ApiDOM tree 1`] = `
(XmlElement
  (MemberElement
    (StringElement)
    (StringElement))
  (MemberElement
    (StringElement)
    (StringElement))
  (MemberElement
    (StringElement)
    (StringElement))
  (MemberElement
    (StringElement)
    (BooleanElement))
  (MemberElement
    (StringElement)
    (BooleanElement)))
`;
